### ISSUE TITLE
Improve the composer swipe down closing logic

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#54270]
Coming from [this thread](https://expensify.slack.com/archives/C05LX9D6E07/p1734434242143269) in quality room.

### Problem

I, as a user, are used to the fact that in every chat app I use I can interactively dismiss a keyboard. That means, when the keyboard in a chat is open to close it I can simply move my finger down the screen and as soon as my finger touches the keyboard, the keyboard is simultaneously swipes down with the movement of my finger (examples of that can be seen in "Messages App (iOS)", "WhatsApp", "Telegram", "Instagram DMs" etc - see the video attached).

In NewDot it seems like I can't dismiss the keyboard in a chat. Once i opened the keyboard there seem to be only two ways to dismiss the keyboard:
- By navigating away from the screen and
- By making a swipe down gesture on the text input explicitly (:warning: Someone from expensify had to educate me about this gesture explicitly. I didn't figure this out on my own!)

Now there are three problems with this:
- Its very none standard to not be able to dismiss the keyboard interactively. Thats a lacking hygiene factor imo (whatsapp quality)
- I never discovered option (2) on my own, thinking I simply couldn't dismiss the keyboard in chat (which was bothering me)
- When you have a message in the composer that exceeds the line limit scrolling through the message might dismiss the keyboard as of functionality (2) which is annoying and none standard (see video in thread)

### Solution
Lets add interactive keyboard dismiss support. Actually, we are already partially doing that over [at this PR](https://github.com/Expensify/App/pull/42143) which we are working on.
In this PR we are basically migrating the chat to react-native-keyboard-controller which will allow us to implement the dismiss gesture very simply and consistently across platforms.

However, its not the goal of that PR to add it, it only adds the "infrastructure". I would like to open a follow up issue to add support for interactive keyboard dismissal once the mentioned PR is merged.

<details><summary>Issue Owner</summary>Current Issue Owner: @chrispader</details>

**Comment by hannomargelo** [comment]
cc @kirillzyusko can you please comment here, as you will be the one working on this most of the times I think (feel free to assign us both Vit!)

**Comment by hannojg** [comment]
cc @kirillzyusko  can you please comment here, as you will be the one working on this most of the times I think (feel free to assign us both Vit!)

**Comment by kirillzyusko** [comment]
Hello, I'm Kiryl from Margelo expert agency and I would like to work on this task 😊

**Comment by mountiny** [comment]
Thanks! 

Can you please write a plan of how to tackle this first, thanks!

**Comment by hannojg** [comment]
The idea is to first get this PR merged (we can put this issue on hold for it), as it restructures the report screen to use `react-native-keyboard-controller`:

- https://github.com/Expensify/App/pull/42143

I think the next step once this is merged would be to build a PoC PR where we implement the interactive keyboard dismissal using RNKC on the report screen.

From this PoC we can decide if it's good to get merged already or should be split up in smaller tasks (thats a bit hard to foresee right now. My gut says/hopes that one PR is enough and the changes won't be too many).

Once we got those landed we could check if there are other product areas that could benefit from improved keyboard handling.

**Comment by CortneyOfstad** [comment]
Hey team! Just a heads up that I will be OOO starting this afternoon (December 20th) and will be returning January 6th. A handful of folks on the BZ team will be online for a few days in between the 25th and the 1st, but we'll be operating with a skeleton crew. If any action is needed from a BZ perspective, please post this issue in #expensify-open-source and someone on the team will jump in. Otherwise I will review when I'm back in office. 

Thank you and Happy Holidays!

**Comment by kirillzyusko** [comment]
I think we should put this issue on hold until https://github.com/Expensify/App/pull/42143 gets merged 👀 

That PR has a fundamental work and it'll be easier to adopt a new logic over a solid fundament rather than begin everything from the scratch 🙂

**Comment by hannojg** [comment]
@MelvinBot still on hold for https://github.com/Expensify/App/pull/54764

**Comment by CortneyOfstad** [comment]
Thanks @hannojg!

**Comment by CortneyOfstad** [comment]
Still on hold!

**Comment by roryabraham** [comment]
Seems like this should be on HOLD for https://github.com/Expensify/App/issues/37203 as well. I have some thoughts/hopes on how we might standardize keyboard handling logic across the app

**Comment by roryabraham** [comment]
I also have a lot of historical context on #54764, so I'm going to assign myself here

**Comment by CortneyOfstad** [comment]
Thanks Rory!

**Comment by CortneyOfstad** [comment]
@roryabraham any update? Looks like the last update on the on-hold issue was about 2 weeks ago. Thanks!

**Comment by roryabraham** [comment]
The most recent update I'm aware of can be found here: https://expensify.slack.com/archives/C035J5C9FAP/p1738949968831959

**Comment by CortneyOfstad** [comment]
Thanks Rory!

Bumped the on-hold PR as it looks like it hasn't had any movement for 3 weeks 👍

**Comment by roryabraham** [comment]
not overdue - PR is moving

**Comment by CortneyOfstad** [comment]
Just a heads up that I am OoO starting this afternoon — March 28 to April 7. If any action is needed via BZ, please reapply the label, otherwise I will review when I get back. Thanks!

**Comment by CortneyOfstad** [comment]
PR is still being drafted

**Comment by CortneyOfstad** [comment]
PR is still being worked on

**Comment by CortneyOfstad** [comment]
PR is still open

**Comment by CortneyOfstad** [comment]
PR is still moving!

**Comment by CortneyOfstad** [comment]
PR has a few more conflicts to resolve (per comment [here](https://github.com/Expensify/App/pull/54764#issuecomment-2902808253)) but looks to be merged shortly! 🎉

**Comment by mountiny** [comment]
PR was merged @kirillzyusko would you be able to continue on this task?

**Comment by CortneyOfstad** [comment]
Bump @kirillzyusko ^^^ thanks!

**Comment by kirillzyusko** [comment]
Thank you @CortneyOfstad for a reminder 🙌 

Yes, since https://github.com/Expensify/App/pull/54764 was merged we can proceed and work on this ticket 👍

**Comment by CortneyOfstad** [comment]
Sounds good! 

Just a heads up that I am heading OoO this afternoon (6/3) to Monday (6/9) so if any action is needed from a BZ perspective, please reapply the label. Otherwise I will get it addressed as soon as I am back. Thanks!

**Comment by CortneyOfstad** [comment]
Hey team — I'm back! @kirillzyusko where are we at with this? Thanks!

**Comment by kirillzyusko** [comment]
@CortneyOfstad @chrispader is going to work on this ticket 👀 

@chrispader can you give an actual status?

**Comment by CortneyOfstad** [comment]
@chrispader — can you provide an update? Thanks!

**Comment by chrispader** [comment]
@CortneyOfstad haven't found time to work on this yet unfortunately. Will try to tackle this at the begin of next week!

**Comment by CortneyOfstad** [comment]
Sounds good – thanks @chrispader!

**Comment by CortneyOfstad** [comment]
Adjusted the date range as it is considered to be a Critical issue 👍

**Comment by wgsquayson** [comment]
I was able to overcome the last issue. Now the chat screen has scroll events are being tracked by the useAnimatedScrollHandler hook(at least on ios for now, since i'm using the ActionSheetAwareScrollView implementation. hopefully nothing else is breaking 😅). 
The video below shows [a known bug of react-native-keyboard-controller](https://github.com/kirillzyusko/react-native-keyboard-controller/issues/594#issuecomment-2373283306), that happens when KeyboardAvoidingView is used with interactive dismissal. Some sort of double scroll happens. It is possible to disable the keyboard avoiding behavior, but the input and chat position will have to be handled manually. This is the next step, for now.

https://github.com/user-attachments/assets/fffa2619-2d7e-43b0-ba07-887edad90366

**Comment by mountiny** [comment]
Thanks, discussing in the thread [here](https://expensify.slack.com/archives/C05LX9D6E07/p1751397605975169?thread_ts=1751357613.249229&cid=C05LX9D6E07)

**Comment by wgsquayson** [comment]
Hey - haven't had time to look into this today but tomorrow i'll get back to it.

**Comment by wgsquayson** [comment]
Today I was investigating the reason why the double scroll behavior previously mentioned was happening - no success fixing it. And I was not able to find a way to properly handle the scrollview/footer positioning yet.
My current approach is trying to apply [this](https://github.com/kirillzyusko/react-native-keyboard-controller/blob/main/example%2Fsrc%2Fscreens%2FExamples%2FInteractiveKeyboardIOS%2Findex.tsx) to the chat, but the double scroll is still an issue.

**Comment by wgsquayson** [comment]
Today I figured what I was doing wrong yesterday - i was missing some content inset/offset animation configs in order to make the scroll position work as expected. Also, some tweaking to make the config that i mentioned on my last message to work on inverted flatlists. It still needs some refinement, I'll have a pair w/ Kiryl from Margelo tomorrow to discuss on it. But it's progressing. 

<video src='https://github.com/user-attachments/assets/9c53cc8c-bf10-414f-9e6f-a0d6ba429c94' />

**Comment by wgsquayson** [comment]
So, today i was able to make it work, it's mostly done on ios. Next steps are
- Figure out why there is a small padding between the keyboard and the input when the keyboard is opened, i'm guessing it has something to do with safe area but i really don't know
- Consider the input height for the gesture to work whenever it grows
- Add a handler for when the user exits from the chat, the keyboard dismisses
- Setting up the behavior for android
- Test across different sized devices
- Probably updating some test cases

<video src='https://github.com/user-attachments/assets/0ae3bf05-e29c-4297-aa4a-fa3c4493bb6e' />

**Comment by mountiny** [comment]
Thanks for the update

**Comment by wgsquayson** [comment]
Hey, still progressing

From my last list, the checked items are

> Figure out why there is a small padding between the keyboard and the input when the keyboard is opened, i'm guessing it has something to do with safe area but i really don't know

It was the bottom spacing that the ScreenWrapper component applies. I had to disable it and handle the bottom padding manually (useSafePaddings hook) in order to mantain the consistency with the current design. One problem happened - notice on the video below how when the keyboard opens, there is a bigger spacing between the composer and the chat. That's because the padding bottom of the chat view now equals the composer height + the bottom padding.  I'm finding a way to properly handle it, until right now what i tried made the text jump.

> Consider the input height for the gesture to work whenever it grows

Done

>Add a handler for when the user exits from the chat, the keyboard dismisses

Done

<video src='https://github.com/user-attachments/assets/111915d1-eb1c-437d-9339-0d4eef9bc4cf' />

**Comment by wgsquayson** [comment]
Hi!
Today i worked on and tested the android implementation - The KeyboardGestureArea component handles it pretty well.
One thing that I noticed is that on iOS the content is not being lifted up from where the scroll position is, and i'll dig into that tomorrow.

**Comment by wgsquayson** [comment]
Still progressing w/ the investigation of my last comment.

**Comment by mountiny** [comment]
@wgsquayson thank you for the updates, is there anyone to tap into for help again?

**Comment by wgsquayson** [comment]
Hey
I'm getting closer to a solution, so far i was able to preserve the scroll position when the keyboard opens/closes but i'm still refining on the case the user drags to close the keyboard. Tomorrow I'll be pairing w/ a teammate.

**Comment by mountiny** [comment]
Ok thank you for the update

**Comment by wgsquayson** [comment]
Hey!
Today I had some issues with the expensify app related to the metro bundler, fast refresh was just not working out of nowhere. So i had to make a clean reinstall of it after trying some fixes. Still progressing with the modal issue. So far, this is the current overall progress on ios.

<video src='https://github.com/user-attachments/assets/e327b55a-1dc3-4f68-82ac-6c4f81edaed6' />

**Comment by mountiny** [comment]
Great, thank you for the update!

**Comment by wgsquayson** [comment]
Hey, good news - i was able to handle the 'modal opens when keyboard is up case'
But, on my saga of finding bugs as i go, another one came up - when the composer is full size, the layout breaks - on iOS, the list offsets are being updated incorrectly, apparently. Looking into solutions.

**Comment by roryabraham** [comment]
@mountiny I believe https://github.com/Expensify/App/pull/65517 is the draft PR in question

**Comment by wgsquayson** [comment]
@mountiny i did some changes to the code today that you might want to generate another build

Today's updates:
- Had to replicate most of the keyboard logic from ReportActionsList to MoneyRequestReportActionsList
- Tested across OS's and corrected some small stuff
- Had to adapt the tests of the useReportUnreadMessageScrollTracking since it uses an animated reaction handler now

I guess the only missing thing is the PR checklist but i'll need to leave now, and I'll fill it on monday.

**Comment by wgsquayson** [comment]
Hello!
I had to remake the onStartReached and onEndReached implementations on the lists that are using the interactive dismissal, since ScrollView does not support it out of the box. I'm still correcting some tests.

**Comment by wgsquayson** [comment]
Sorry, I forgot to write yesterday. But I'm investigating why onEndReached and onStartReached are not working.

**Comment by wgsquayson** [comment]
Hey
I haven't had much time to put into this today, but I was came to some conclusions on the investigation
Basically it's because of the onScroll function being handled by reanimated's useAnimatedScrollHandler. The onScroll implementation of virtualized lists is the one that handles the edges calculation and the calling of onEndReached and onStartReached. So, the next step is finding a way to fix this limitation, because just implementing it might break other stuff.

**Comment by wgsquayson** [comment]
Hello!

Today I met w/ Chris, and I moved forward with the solution of creating a KeyboardDismissalFlatList component (that points to a custom implementation of the Animated.FlatList, pretty much the same stuff but accepts a CellRendererComponent) and context (basically a copy of the useKeyboardAnimation handler) in order to fix the scroll problem and handle the keyboard animation. Now the onStartReached/onEndReached work. More details are on [this commit](https://github.com/Expensify/App/pull/65517/commits/caee34bdcd0b9c5350a240d7e44a7aa43c4fff36).

I should point out that after rebasing to expensify's main the ActionSheetAwareScrollView is not working properly (at least not on iOS) anymore and it's not related to my PR, i switched to expensify's main and a long pressed message is not going above the action sheet.

**Comment by wgsquayson** [comment]
Hey, updates from yesterday and today
- I found an edge case with the offline indicator on iOS, that the indicator was not coming up w/ the keyboard, already fixed it
- I moved the useKeyboardDismissibleFlatListContext  hook to a separate file to properly mock it
- I noticed some tests are failing but idk if they are related to my changes, since on main they also break

**Comment by CortneyOfstad** [comment]
Thanks @wgsquayson!

**Comment by wgsquayson** [comment]
Hello! Rajat did some testing, came back with some bugs/code comments and i just updated the PR and left some comments as well.

**Comment by CortneyOfstad** [comment]
PR continues to move 🎉

**Comment by CortneyOfstad** [comment]
PR hit a big of a snag with a bug (mentioned [here](https://github.com/Expensify/App/pull/65517#issuecomment-3267093052)) but has already been fixed 👍 PR Is ready to be tested again

**Comment by CortneyOfstad** [comment]
Original PR has been split into two smaller PRs – https://github.com/Expensify/App/pull/71438 & https://github.com/Expensify/App/pull/71440

**Comment by CortneyOfstad** [comment]
Hey @parasharrajat — any update? Thanks!

**Comment by CortneyOfstad** [comment]
@parasharrajat Bump ^^^^

**Comment by parasharrajat** [comment]
First PR is reviewed. It will be merged soon.

**Comment by chrispader** [comment]
> I think it would be appropriate to make this a weekly.
> 
> [@chrispader](https://github.com/chrispader) and [@parasharrajat](https://github.com/parasharrajat) would likely be best positioned to provide an update on next steps, but I know [@chrispader](https://github.com/chrispader) is doing a sprint this week and will be otherwise occupied.

Yes, as @roryabraham mentioned i'm on a feature sprint this week, so i'll hopefully be able to continue working on this after the christmas holidays.

Now that we've added and updated the components for interactive keyboard dismissal in https://github.com/Expensify/App/pull/71438, i'll continue on https://github.com/Expensify/App/pull/71440 which enables the new logic in the lists.

**Comment by CortneyOfstad** [comment]
Not overdue!

**Comment by roryabraham** [comment]
I'm not sure - @chrispader was going to handle next steps but I know he has a lot on his plate.

@chrispader can you write down the next steps to finish this issue out? That way maybe we can hand it off?

**Comment by mountiny** [comment]
RNKC prs merged

**Comment by chrispader** [comment]
@mountiny and i decided to put this aside for now and i'll continue working on it once i've finished other tasks with higher priority, since this is an UI/UX improvement and not super urgent.

Technically, @kirillzyusko and i have to fix one remaining issue that we have with `react-native-keyboard-controller`. I'll check soon whether we can find a quick fix for this and upstream a fix.

**Comment by chrispader** [comment]
Since we're migrating all the lists in the app to LegendList, this will make interactive keyboard dismissal way easier and less complex, because LegendList integrates with RNKC out-of-the-box with `KeyboardAwareLegendList`. I'm working in person with @kirillzyusko this week on getting this to work! 🙌🏼

Then we'll just need to wait for PRs 2-4 in https://github.com/Expensify/App/issues/98994 to be merged.
